vip_hci/hci_dataset.py

```diff
@@ -789,16 +793,101 @@ def inject_companions(self, flux, rad_dists, n_branches=1, theta=0,
             if self.wavelengths is None:
                 raise ValueError('The wavelengths vector has not been set')
 
-        self.cube, yx = cube_inject_companions(self.cube, self.psfn,
-                                               self.angles, flux, self.px_scale,
-                                               rad_dists, n_branches, theta,
-                                               imlib, interpolation,
-                                               full_output=True,
-                                               verbose=verbose)
+        self.cube, yx = cube_inject_companions(
+            self.cube, self.psfn, self.angles, flux, self.px_scale,
+            rad_dists, n_branches, theta, imlib, interpolation,
+            full_output=True, verbose=verbose
+        )
+
+        if not hasattr(self, "injections_yx"):
+            self.injections_yx = []
+
+        self.injections_yx += yx
+
+        if verbose:
+            print("Coordinates of the injections stored in self.injections_yx")
 
         if full_output:
             return yx
 
+    def copies_with_injections(self, n_copies, inrad=8, outrad=12,
+                               dist_flux=("uniform", 2, 500)):
+        """
+        Create copies of this dataset, containing different random injections.
+
+        Parameters
+        ----------
+        n_copies : int
+            This is the number of 'cube copies' returned.
+        inrad,outrad : float
+            Inner and outer radius of the injections. The actual injection
+            position is chosen randomly.
+        dist_flux : tuple('method', *params)
+            Tuple describing the flux selection. Method can be a function, the
+            ``*params`` are passed to it. Method can also be a string, for a
+            pre-defined random function:
+
+                ``("skewnormal", skew, mean, var)``
+                    uses scipy.stats.skewnorm.rvs
+                ``("uniform", low, high)``
+                    uses np.random.uniform
+                ``("normal", loc, scale)``
+                    uses np.random.normal
+
+        Returns
+        -------
+        fake_datasets : list of HCIDataset
+            List containing ``n_copies`` fake datasets.
+
+        """
+
+        fake_datasets = []
```

Comment:
I'm in favor of this new method. The only problem I see is the memory management. Let's say we have a 10 Gb datasets, and we request 100 copies... I'd suggest to check the available memory as it's done in vip_hci.pca.pca (full-frame PCA)

Reply:
I just added the memory checking using `conf.mem.check_enough_memory`. However, maybe we could think of this from another perspective: instead of storing all the cubes/datasets in a big list, we could also separate the problem in two steps:

1. Create the information required for the injection (position, flux, ...) and store it in a variable. That could also be stored to disk, to reproduce the injections later.
2. Based on that information, use a generator/iterator, so that only the cube, which is currently processed, is stored in memory.

in code:

``` python
ds = HCIDataset(...)
inj_info = make_injections(angles, n_copies, dist_flux=...)
for dataset in ds.generate_copies_with_injections(inj_info):
    # just this current dataset ist stored in memory, and
    # deleted on the next iteration element
```

I'll think about a simple way to do so.